homeassistant/scripts/check_config.py

```diff
@@ -0,0 +1,172 @@
+              color('not used', c=Fore.RED)+')')
+        # Python 3.5 gets a recursive, but not in 3.4
+        for yfn in sorted(glob(os.path.join(config_dir, '*.yaml')) +
+                          glob(os.path.join(config_dir, '*/*.yaml'))):
+            the_color = '' if yfn in res['yaml_files'] else Fore.RED
+            print(color('-', yfn, c=the_color))
+
+    # pylint: disable=too-many-nested-blocks
+    if SHOW_FULL_CONFIG:
+        re_keys = re.compile(r"'([^']+)':")
+
+        def indent(item, indent0, indent1):
+            """Return indented text."""
+            return re_keys.sub(color(r'\1')+':',
+                               fill(pprint.pformat(item),
+                                    initial_indent=indent0,
+                                    subsequent_indent=indent1))
+
+        def printlist(obj, level=0, pref=''):
+            """Printlist."""
+            indent1 = level * ' '
+            indent0 = pref.rjust(level)
+            if isinstance(obj, list):
+                for ob0 in obj:
+                    printlist(ob0, level+2, '- ')
+            # elif isinstance(obj, dict):
+            #    for key0, val0 in obj.items():
+            #        printlist(val0, indent00 + key0 + ':' , indent1)
+            else:
+                print(indent(obj, indent0, indent1))
+
+        components = deep_convert_dict(res['components'])
+        for name, config in components:
+            if config is None:
+                continue
+            print(color(name + ':'))
+            printlist(config, 2, '')
+
+    return 0
+
+
+def check(config_path):
+    """Perform a check by mocking hass load functions."""
+    # List of components/platforms
+    components = []
+    yaml_files = {}
+    secrets = {}
+
+    original_get_component = loader.get_component
+    original_load_yaml = yaml.load_yaml
+
+    def mock_load_yaml(filename):
+        """Mock hass.util.load_yaml to save config files."""
+        yaml_files[filename] = True
+        return original_load_yaml(filename)
+
+    def mock_get_component(comp_name):
+        """Mock hass.loader.get_component to replace setup & setup_platform."""
+        def mock_setup(*kwargs):
+            """Mock setup, only record the component name & config."""
+            # pylint: disable=cell-var-from-loop
+            components.append((comp_name, kwargs[1].get(comp_name)))
+            return True
+
+        module = original_get_component(comp_name)
+        # Test if platform/component. Also: '.' in comp_name = platform
+        if hasattr(module, 'setup'):
+            module.setup = mock_setup
+        if hasattr(module, 'setup_platform'):
+            module.setup_platform = mock_setup
+
+        module.setup_platform = mock_setup
+        return module
+
+    @patch("homeassistant.util.yaml.load_yaml",
+           side_effect=mock_load_yaml)
+    @patch("homeassistant.loader.get_component",
+           side_effect=mock_get_component)
+    def load_all_the_config(*mocks):
```

Comment:
Just name the mocks, you know exactly what will be passed in.

Actually, you can use `patch` as a context manager instead of a decorator and simplify this code.

```python
with patch('homeassistant.util.yaml.load_yaml',
           side_effect=mock_load_yaml) as m_load_yaml:
    …
```